### ISSUE TITLE
Segfault in Unity 2018.3.8 compiled on Android using IL2CPP

### ISSUE THREAD

**Issue by jweissBHG** [MiloszKrajewski/K4os.Compression.LZ4#19]
When running a Unity 2018.3.8 app on Android compiled with IL2CPP, I'm getting a SIGSEGV when calling `LZ4Codec.Decode()`. The encoded data is coming from a server running .Net Core 2.1 on Linux, encoded with `LZ4Codec.Encode()`. Everything works fine when running Unity in editor, I'm experience the segfault only on Android devices. Do you have any idea what could be going wrong in this build scenario?

**Comment by jweissBHG** [comment]
Did some more debugging today and narrowed it down to this line:

https://github.com/MiloszKrajewski/K4os.Compression.LZ4/blob/d7bd4fe37f488c9f0f56251cf1de1217eb3b4090/src/K4os.Compression.LZ4/Engine/LZ4_xx.cs#L257

```
Mem.WildCopy(op, ip, cpy);
```

Going to be getting more info soon

**Comment by jweissBHG** [comment]
Not sure why, but it seems like the `Mem.cs` methods that utilize `ulong*` are the ones triggering the segfaults. If I switch it to `int*` math instead of `ulong*` math (and all the increments by 8 to 4 instead), I can get rid of the segfault. Not too sure why this is happening, but I have worked past it

**Comment by jweissBHG** [comment]
Current thinking is that this is a Unity IL2CPP bug that has to do with `ulong*` pointers, am going to try to make a basic project for reproducing the issue when I have time.

**Comment by MiloszKrajewski** [comment]
My bet would be on unaligned memory acces

**Comment by MiloszKrajewski** [comment]
Do I need to switch to 32 bit on 32-bit processors or it will be fixed in mono/unity?
Can you update me on this? (I assume you are monitoring this issue)

**Comment by capyvara** [comment]
@MiloszKrajewski I did a similar fix on neuecc/MessagePack-CSharp#463 above, I think it's unlikely to be fixed on unity side, so we'll need to use a 32bit version of Mem.WildCopy(), at least on 32 bit processors, we can branch in runtime by using Environment.Is32BitProcess() or IntPtr.Size == 4.

**Comment by MiloszKrajewski** [comment]
I just hoped for not branching at runtime. WildCopy is meant to be fast, and branching kills it. It is also inlined method so when it becomes quite big it is killing processor cache.
It is not WildCopy though. It is Fill, Zero, Copy8, Copy16, and Copy18.
Argh. Need to think about it.

**Comment by capyvara** [comment]
I think the idea is not to branch in the Mem methods themselves, but as early as possible. If you check on neuecc implementation there branches for 4 permutations decided right after (32 or 64 bits, < 64kb or > 64kb input length).

However it will require some code duplication, depending on your C# minimum version (7.3+) we could somehow use generics with the unmanaged type constraint to reduce that.

**Comment by jweissBHG** [comment]
I have not had time to investigate further -- I'm also assuming this is something that Unity won't fix, at least not until someone proves that it's their fault and submits a minimal repro project to them (something I have not yet done).

**Comment by MiloszKrajewski** [comment]
I hope I've addressed 32/64/alignment problem in 1.2.1-beta. 
I've rested in on .NET Core 2.2 @ Rasperry Pi 3 and it works fine, but It would be helpful if someone could test on Mono / Android.

**Comment by LittleCodingFox** [comment]
I just tested it and I no longer have the crash. Thanks for releasing a fix!

**Comment by MiloszKrajewski** [comment]
@LittleCodingFox What platform did you test? Unity? Xamarin?

**Comment by LittleCodingFox** [comment]
Unity Android 2019.3 IL2CPP :)

**Comment by MiloszKrajewski** [comment]
Thanks. I'll close the issue.

**Comment by pepoon** [comment]
First of all thanks to MiloszKrajewski for this amazing library.
I tested that on Samsung Galaxy s5(Armvv7) and it crashes..

What I'm doing is:
- Get the repository in VS using nuget. 
- Import the libs & dependences to UNITY.

But once I Deploy to Armv7 IL2CPP it crashes,
Then I checked the repository and i found these lines:

namespace K4os.Compression.LZ4.Engine
--
  | {
  | #if BIT32
  | internal unsafe class LZ4_32: LZ4_xx
  | {
  | ...
  | #else
  | internal unsafe class LZ4_64: LZ4_xx
  | { }

That means to make it work on Armv7 devices, I need to pull the solution and define/or not the #32BIT processor, get the .dlls and use them in UNITY to make it work on Armv7?
Are these the steps?
Thanks.

**Comment by MiloszKrajewski** [comment]
No you don't. It should work. Can you send me minimal working (I mean crashing) project? 
Unity or Visual Studio/Xamarin?

**Comment by pepoon** [comment]
yea in 3h ill make one.
thanks

**Comment by MiloszKrajewski** [comment]
Can you try setting:
```
LZ4Codec.Enforce32 = true;
```
on startup?

**Comment by pepoon** [comment]
Yea, It crashes aswell,
Preparing a light project now..
Let's hope it is my mistake.

Missatge de Milosz Krajewski <notifications@github.com> del dia dj., 28 de
maig 2020 a les 16:22:

**Comment by MiloszKrajewski** [comment]
> [...] Let's hope it is my mistake. 

I don't thinks so. If it was throwing some "normal" exception or not compressing random data (which by definition cannot be compressed) them maybe but proper crash is hardware.

And often I have no idea what is actually wrong (last time if was 64-bit unaligned writes on 32-bit platform).

**Comment by pepoon** [comment]
I have good news, it is not happening on an empty project with "Hello
world" text.
I'll dig more.
Sorry for any inconvenience caused.


Missatge de Milosz Krajewski <notifications@github.com> del dia dj., 28 de
maig 2020 a les 18:48:

**Comment by pepoon** [comment]
I finally could reproduce the bug, Here are the logcat logs and an apk with
IL2CPP armv7 & armb64.
Pushing the Unity project as well later.
It is trying to decompress a file. On armv64 it works while on armv7
crashes.
 arm7-logs.txt
<https://drive.google.com/file/d/1q2_uWqv6Q-xtjebVzAm_a4EWCL9GR6Cm/view?usp=drive_web>
 arm64-logs.txt
<https://drive.google.com/file/d/1_g9-JyQ16e0HrKuDlMoNA8yrzlgozd-2/view?usp=drive_web>
 lz4-test2-arm7-64-.apk
<https://drive.google.com/file/d/1OLa81yuTJKstwAIyCofYuWfSh24SSjZ7/view?usp=drive_web>

**Comment by LittleCodingFox** [comment]
The google drive links don't have permission, you need to enable link sharing

**Comment by pepoon** [comment]
> The google drive links don't have permission, you need to enable link sharing

Sorry, I think now i did it.

**Comment by LittleCodingFox** [comment]
yep, works now :)

**Comment by pepoon** [comment]
[image: image.png]
 Lz4_test.zip
<https://drive.google.com/file/d/1Su7TveX2KjeyvcNodrfAV_KoZAZGYgWD/view?usp=drive_web>
[image: image.png]
Unity project Attached:
https://drive.google.com/file/d/1Su7TveX2KjeyvcNodrfAV_KoZAZGYgWD/view?usp=sharing

Test is compressing a file then decompressing.
Attached the full unity project.
Unity Version: 2019.3.14.f1



Missatge de Josep Maria Calull Mercadé <calullmercade@gmail.com> del dia
dj., 28 de maig 2020 a les 19:01:

**Comment by pepoon** [comment]
Hey Mikos, Fowarded unity response related to the bug lz4 arm7 crash.

thanks

---------- Forwarded message ---------
From: Unity QA <bugs@unity3d.com>
Date: Thu, Jul 23, 2020, 3:17 PM
Subject: Re: (Case 1251505) Armv7 vs Armv64 IL2CPP crashes on
decompres/compress assetbundle using custom compression
To: Josep Maria Calull Mercadé <calullmercade@gmail.com>


Josep,

Thanks for reporting this issue. It looks like the code here will do an
unaligned memory access, which is not allowed on armv7, hence the crash.
Mono works in this case because it generates code that is less efficient
than IL2CPP, and does only aligned memory access.

With IL2CPP, we have chosen to convert the C# code as-is, so that the
generated code will do unaligned access if the C# code does.

I would recommend following up with the library developer - although
looking at the Github issue thread, that might already be the case.

Please let me know if you have any questions about this issue. Thanks again!

Josh Peterson
Senior Software Developer

-- 
Thank you for submitting your support request! In the meantime I would
suggest checking out http://answers.unity3d.com/
and our Forums at http://forum.unity3d.com/
for additional help.

Track the status of your ticket here:
https://fogbugz.unity3d.com/default.asp?1251505_l59gi75pmf7g5vch
(Please do not share this url with anyone else.  Doing so will give them
access to all information you share with us.  If you post this url on
forums or blogs we cannot prevent others from viewing all your support
cases and bug reports.)



-----Original Message-----
From: "Josep Maria Calull Mercadé" <calullmercade@gmail.com>
Reply-To: "Josep Maria Calull Mercadé" <calullmercade@gmail.com>
Date: Wed, 3 Jun 2020 10:21:54 +0200
To: "Monika Pociute" <bugs@unity3d.com>
Subject: Re: (Case 1251505) Armv7 vs Armv64 IL2CPP crashes on
decompres/compress assetbundle using custom compression

>Thank you for your response.
>samsung galaxy s5 it crashes(armv7)
>while armv64 device will not.
>i just build and apk+il2cpp and play.
>
>mono not crashing.
>
>On Wed, Jun 3, 2020, 9:48 AM Monika Pociute <bugs@unity3d.com> wrote:
>
>> Hi,
>>
>> Thank you for contacting Unity about your issue.
>>
>> Unfortunately, we are not able to reproduce it. Are you still
experiencing
>> this problem or was it a one time issue?
>> If you're still facing this problem, could you please provide us with
more
>> information - the exact steps needed to reproduce on our side?
>>
>> Thanks,
>> Monika
>> Customer QA Team
>>

**Comment by MiloszKrajewski** [comment]
As I said [before](https://github.com/MiloszKrajewski/K4os.Compression.LZ4/issues/19#issuecomment-481181587) it is unaligned memory access.
I was hoping that using stricly 32bit mode will address the problem, but it seems it did not.
I'n my opinion it is very unclear how and if ARM handles it. Sometimes (https://medium.com/@iLevex/the-curious-case-of-unaligned-access-on-arm-5dd0ebe24965#:~:text=Hence%2C%20unaligned%20access%20simply%20means,byte%20alignment%20for%20optimal%20performance.) it says it does.

I'm sorry that you (or anyone) is affected by this bug but there is no east fix for it.

**Comment by MiloszKrajewski** [comment]
This issue has been revied as #40.
I'll need some help testing it as I don't have access to devices which are actually crashing.

@pepoon @LittleCodingFox
